moosez/image_conversion.py

```diff
@@ -36,3 +38,79 @@ def dcm2nii(dicom_dir: str) -> None:
     subprocess.run(cmd, shell=True, capture_output=True)
     spinner.succeed(text=f"Converted DICOM images in {dicom_dir} to NIFTI")
     logging.info("Conversion completed successfully")
+
+    
+def read_dicom_folder(folder_path: str) -> SimpleITK.Image:
+    """
+    Reads a folder of DICOM files and returns the image
+
+    :param folder_path: str, Directory to get DICOM files from
+    """
+    reader = SimpleITK.ImageSeriesReader()
+    dicom_names = reader.GetGDCMSeriesFileNames(folder_path)
+    reader.SetFileNames(dicom_names)
+
+    dicom_image = reader.Execute()
+    return dicom_image
+
+
+def anyformat2nii(input_path: str, output_directory: str = None) -> None:
+    """
+    Converts any image format known to ITK to NIFTI
+
+    :param input_path: str, Directory OR filename to convert to nii.gz
+    :param output_directory: str, optional output directory to write the image to.
+    """
+    if os.path.isdir(input_path):
+        image_probe = os.listdir(input_path)[0]
+        if image_probe.endswith('IMA') or image_probe.endswith('dcm'):
+            # Get a few infos from the probed file's basename
+            image_probe_basename = os.path.basename(image_probe)
+            input_extension = image_probe_basename[image_probe_basename.rfind('.') + 1:]
+            image_probe_stem = os.path.basename(input_path)
+            logging.info(f"Converting {input_extension} image to NIFTI")
+
+            # Reading the dicom directory
+            output_image = read_dicom_folder(input_path)
+
+            # The new image stem is generated here
+            output_image_basename = image_probe_stem + '.nii.gz'
+        else:
+            logging.info(f"Conversion ERROR: The specified directory does not contain a dicom or IMA file.")
+            return
+
+    elif os.path.isfile(input_path):
+        if input_path.endswith('.nii.gz'):
+            logging.info(f"A compressed NIFTI file was provided. No conversion will follow.")
+        else:
+            # Get a few infos from the file basename
+            input_image_basename = os.path.basename(input_path)
+            input_extension = input_image_basename[input_image_basename.rfind('.') + 1:]
+            input_image_stem = input_image_basename[:input_image_basename.rfind('.')]
+            logging.info(f"Converting {input_extension} image to NIFTI")
+
+            # Determining if the file is a DICOM file or not, so we can search deeper
```

Comment:
Hi, I am not able to understand this code bit here. What are we trying to do here?

Reply:
input_image_stem = input_image_basename[:input_image_basename.rfind('.')] gets the file name without the extension, so basically, a substring of the basename spanning from the first position to the index where the dot for the file extension was found. This is needed to generate the new filename and to append it to .nii.gz

Reply:
If a path to a file is provided, only the second check is done, as the first is skipped. If the provided `input_path` leads to a dicom image, it automatically searches for all other dicom images in the folder to build a volume instead of reading only one slice.

Reply:
ok good, I will push it in.